### PR TITLE
refactor(router/atc): cleanup split routes and services

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Just a slightly tighter code and reusing table.

### Checklist

- [ ] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

This was originally proposed at #12901 for inclusion in #12875.

### REBASE AND MERGE

### CODE REVIEW

File: kong/router/transform.lua [hunk 1]
```diff
@@ -733,42 +733,36 @@ local function split_routes_and_services_by_path(routes_and_services)
...
-
       -- create a new route from the original route
       route.original_route = original_route
-      route.paths = paths
-      route.id = uuid_generator(original_route_id .. "#" .. tostring(key))
+      route.paths = grouped_paths[j]
+      route.id = uuid_generator(original_route_id .. "#" .. j)
 
-      local cloned_route_and_service = {
+      local index = j == 1 and i or count + 1
```
Comment (chronolaw) [inline kong/router/transform.lua:756]: We should add some comments for this appending logic, it is a little hard to understand.

File: kong/router/transform.lua [hunk 2]
```diff
@@ -696,32 +696,32 @@ end
...
-  return is_regex_magic(path) or #path
-end
-
+-- group array-like table 'paths' to regex and length groups
+local path_groups = {}
+local path_groups_map = {}
+local function group_by_regex_or_length(paths)
+  tb_clear(path_groups)
+  tb_clear(path_groups_map)
+  local path_groups_count = 0
```
Comment (chronolaw) [inline kong/router/transform.lua:705]: Add a blank line for readability.

File: kong/router/transform.lua [hunk 3]
```diff
@@ -696,32 +696,32 @@ end
...
 
 
-local function sort_by_regex_or_length(path)
-  return is_regex_magic(path) or #path
-end
-
+-- group array-like table 'paths' to regex and length groups
+local path_groups = {}
+local path_groups_map = {}
+local function group_by_regex_or_length(paths)
```
Comment (chronolaw) [inline kong/router/transform.lua:702]: This function is not easy to understand, could we add some comments or examples for it?

**Comment by chronolaw** [comment]
This PR has almost the same logic as the original code, but it also introduces two extra tables, the complex data structure make the code hard to understand, we should add more comments for future's maintenance.

**Comment by bungle** [comment]
> This PR has almost the same logic as the original code, but it also introduces two extra tables, the complex data structure make the code hard to understand, we should add more comments for future's maintenance.

It has less code than before and is also more performant, does less new table creations and clones, does not use pairs for looping.

File: kong/router/transform.lua (hunk 2 above)
Comment (bungle) [reply kong/router/transform.lua:685]: @chronolaw blank like added.

File: kong/router/transform.lua (hunk 3 above)
Comment (bungle) [reply kong/router/transform.lua:682]: I added documentation about it.

**Comment by bungle** [comment]
> This PR has almost the same logic as the original code, but it also introduces two extra tables, the complex data structure make the code hard to understand, we should add more comments for future's maintenance.

@chronolaw comments were added. I also added another commit to clear the transformation functions.

File: kong/router/transform.lua [hunk 4]
```diff
@@ -696,34 +675,37 @@ end
...
     else
-      result[key] = { value }
+      path_groups_count = path_groups_count + 1
+      path_groups_map[k] = path_groups_count
+      path_groups[path_groups_count] = { path }
     end
   end
-  return result
+
+  return path_groups_count, path_groups
```
Comment (chronolaw) [inline kong/router/transform.lua:699]: We have the different names where it be called (grouped_paths_count, grouped_paths), could we rename them to be more understandable?

Comment (bungle) [reply kong/router/transform.lua:699]: @chronolaw then the local variables would have same name as module level variables, which is just confusing (aka reusing module level variable name as local variable name).

**Review by chronolaw — APPROVED** [review]
Great job, thank you Aapo.

**Review by samugi — APPROVED** [review]

**Pull request merged**